File: monkey/gh_writeback.py
"""Write-back to GitHub through the token proxy + pre-merge gates.

This is the part that turns an adapter Outcome into an actual GitHub action.
Nothing here ever touches GITHUB_TOKEN (gh-proxy does). Gates run before opening
a PR to keep the bot from junk-merging its own changes. Errors here never
escalate the token.
"""

from __future__ import annotations

import re
import subprocess
from pathlib import Path

from .config import get_settings
from .host_tools import GHProxy


class GateError(Exception):
    pass


async def write_back(
    outcome,
    topic: str,
    *,
    owner: str,
    repo: str,
    number: int,
    store,
    worktree: Path,
) -> dict:
    """Post the appropriate GitHub action based on the task kind / outcome.

    `topic` is the task kind from dispatch (fix | comment | answer | invalid).
    """
    settings = get_settings()
    proxy = GHProxy(
        settings.gh_proxy_url,
        settings.gh_proxy_hmac_key,
        store,
        owner,
        repo,
        number,
    )

    if topic in ("fix",):
        return await _open_pr_if_gated(
            proxy, outcome, owner, repo, number, worktree=worktree
        )
    if topic == "answer":
        # question: one comment, no PR.
        body = outcome.comment or outcome.summary
        await proxy.add_issue_comment(_ensure_comment(body))
        return {"action": "comment", "kind": "answer"}
    if topic == "comment":
        body = outcome.comment or outcome.summary
        await proxy.add_issue_comment(_ensure_comment(body))
        return {"action": "comment", "kind": "comment"}
    if topic == "invalid":
        body = outcome.comment or outcome.summary
        await proxy.add_issue_comment(_ensure_comment(body))
        return {"action": "comment", "kind": "invalid"}

    return {"action": "none", "kind": topic}


async def _open_pr_if_gated(
    proxy: GHProxy,
    outcome,
    owner: str,
    repo: str,
    number: int,
    *,
    worktree: Path,
) -> dict:
    """Open a PR only if the pre-merge gates pass and the body is well-formed."""
    body = outcome.pr_body or outcome.summary
    if not _has_required_headers(body, number):
        # No PR without the required structure - fall back to a comment so the
        # human can see the agent's reasoning.
        await proxy.add_issue_comment(_ensure_comment(body))
        return {"action": "comment_fallback", "reason": "missing_required_headers"}

    branch = getattr(outcome, "branch", "") or ""
    if not branch:
        # No PR without a real branch: the head ref must exist on the remote.
        await proxy.add_issue_comment(_ensure_comment(body))
        return {"action": "comment_fallback", "reason": "missing_branch"}

    # Push the worktree branch so the head ref exists on the remote before
    # GitHub will accept the PR. The token never leaves gh-proxy.
    await proxy.push(worktree, branch)

    pr = await proxy.open_pull_request(
        {
            "title": outcome.summary.split("\n")[0][:120],
            "head": branch,
            "base": "main",
            "body": body,
        }
    )
    return {"action": "open_pr", "pr": pr}


# ---------------------------------------------------------------------------
# Gates
# ---------------------------------------------------------------------------


def gate_pre_push(worktree: Path, branch: str) -> None:
    """Pre-push gates: branch matches, clean tree, commits carry author."""
    _run(["git", "-C", str(worktree), "diff", "--quiet"])
    status = _run(["git", "-C", str(worktree), "status", "--porcelain"])
    if status.strip():
        raise GateError("working tree not clean")

    author_matches = _run(
        ["git", "-C", str(worktree), "log", "-1", "--format=%an <%ae>"]
    ).strip()
    if not author_matches:
        raise GateError("no author on HEAD commit")


def gate_pre_pr(worktree: Path, test_command: list[str] | None = None) -> None:
    """Pre-PR gates: run the repo's checks (fix/check/test if configured)."""
    if test_command:
        _run(test_command)


def _has_required_headers(body: str, number: int) -> bool:
    sections = ("## Repro", "## Cause", "## Fix", "## Verification")
    if not all(s in body for s in sections):
        return False
    # Require a "Fixes #N" / "Closes #N" / "Resolves #N" reference that points at
    # the issue being worked on (#number), not any other issue.
    return bool(re.search(rf"(?:Fixes|Closes|Resolves)\s+#{int(number)}\b", body))


def _ensure_comment(text: str) -> str:
    text = (text or "").strip()
    return text if text else "_no response produced_"


def _run(cmd: list[str]) -> str:
    try:
        proc = subprocess.run(cmd, check=True, capture_output=True, text=True)
    except subprocess.CalledProcessError as exc:
        raise GateError(f"gate command failed: {' '.join(cmd)}\n{exc.stderr}") from exc
    return proc.stdout

File: crates/monkey-github/src/gh_writeback.rs
use regex::Regex;
use serde_json::{Value, json};
use std::path::Path;

use crate::host_tools::GHProxy;
use monkey_core::config::Settings;
use monkey_core::db::Store;
use monkey_core::dispatch::TaskKind;
use monkey_engine::adapters::Outcome;
use monkey_engine::adapters::pi::REPORT_SECTIONS;

/// Identifies the GitHub issue a write-back targets. Grouping the three
/// fields keeps them from drifting apart across function signatures.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct RepoRef {
    pub owner: String,
    pub repo: String,
    pub number: i64,
}

impl RepoRef {
    pub fn slug(&self) -> String {
        format!("{}/{}", self.owner, self.repo)
    }
}

pub async fn write_back(
    outcome: &Outcome,
    kind: TaskKind,
    repo_ref: &RepoRef,
    store: &Store,
    worktree: &Path,
    settings: &Settings,
) -> Result<Value, String> {
    let proxy = GHProxy::new(
        &settings.gh_proxy_url,
        &settings.gh_proxy_hmac_key,
        store.clone(),
        repo_ref,
    );

    match kind {
        TaskKind::Fix => open_pr_if_gated(&proxy, outcome, repo_ref, worktree).await,
        TaskKind::Answer | TaskKind::Comment | TaskKind::Invalid => {
            let body = if !outcome.comment.is_empty() {
                &outcome.comment
            } else {
                &outcome.summary
            };
            proxy.add_issue_comment(&ensure_comment(body)).await?;
            Ok(json!({ "action": "comment", "kind": kind }))
        }
        TaskKind::Skip => Ok(json!({ "action": "none", "kind": kind })),
    }
}

pub async fn open_pr_if_gated(
    proxy: &GHProxy,
    outcome: &Outcome,
    repo_ref: &RepoRef,
    worktree: &Path,
) -> Result<Value, String> {
    let body = if !outcome.pr_body.is_empty() {
        &outcome.pr_body
    } else {
        &outcome.summary
    };

    if !has_required_headers(body, repo_ref.number) {
        proxy.add_issue_comment(&ensure_comment(body)).await?;
        return Ok(json!({
            "action": "comment_fallback",
            "reason": "missing_required_headers"
        }));
    }

    let branch = &outcome.branch;
    if branch.is_empty() {
        proxy.add_issue_comment(&ensure_comment(body)).await?;
        return Ok(json!({
            "action": "comment_fallback",
            "reason": "missing_branch"
        }));
    }

    // Push the worktree branch so the head ref exists on remote
    proxy.push(worktree, branch).await?;

    let first_line = outcome.summary.lines().next().unwrap_or("");
    let title = truncate_utf8(first_line, 120);

    let pr = proxy
        .open_pull_request(json!({
            "title": title,
            "head": branch,
            "base": "main",
            "body": body
        }))
        .await?;

    Ok(json!({ "action": "open_pr", "pr": pr }))
}

pub fn has_required_headers(body: &str, number: i64) -> bool {
    if !REPORT_SECTIONS.iter().all(|&s| body.contains(s)) {
        return false;
    }

    let re_str = format!(r"(?i)(?:Fixes|Closes|Resolves)\s+#{}\b", number);
    if let Ok(re) = Regex::new(&re_str) {
        re.is_match(body)
    } else {
        false
    }
}

pub fn ensure_comment(text: &str) -> String {
    let trimmed = text.trim();
    if trimmed.is_empty() {
        "_no response produced_".to_string()
    } else {
        trimmed.to_string()
    }
}

fn truncate_utf8(text: &str, max_bytes: usize) -> String {
    if text.len() <= max_bytes {
        return text.to_string();
    }

    text.char_indices()
        .take_while(|(index, character)| index + character.len_utf8() <= max_bytes)
        .map(|(_, character)| character)
        .collect()
}
